### PR TITLE
[RFC] Windows: Fix api/vim_spec.lua functional test

### PR DESCRIPTION
**Pull request by sethjackson**
Cherry picked from #810. From equalsraf@80b6abc.

### CODE REVIEW

**Comment by sethjackson** [comment]
Anyone have an idea why this test is failing now? It's failing locally too. I'm sure I'm missing something obvious.

`if has("win32")` in my `nvim` on Unix evaluates to false.

cc @equalsraf @justinmk

**Comment by ZyX-I** [comment]
@sethjackson `has('win32')` returns a number. Numbers are true in lua.

**Comment by sethjackson** [comment]
TIL

We will need to fix [this](https://github.com/neovim/neovim/commit/091e885d441d625fe70c5f05ea2d9a28639ed2e7) then.

**Comment by ZyX-I** [comment]
@sethjackson You need to first make this test fail then. If test contains such bug, but does not fail then there is obviously some other bug inside.

**Comment by sethjackson** [comment]
Is `call serverstart('\\\\.\\pipe\\test_server_pipe')` valid on Unix or no?

**Comment by ZyX-I** [comment]
@sethjackson It may be, but I cannot say what exactly it does. Note that on unix there can exist file named literally `\\.\pipe\test_server_pipe`.

**Comment by sethjackson** [comment]
Right that's why I'm assuming it works since the equality check is messed up.
Although I wouldn't necessarily assume that `\` would work as a path separator character on Unix systems unless `nvim` is handling that on the fly.

I will check.

Regardless I think the equality checks in the tests are fixed now.

**Comment by ZyX-I** [comment]
@sethjackson By the way, socket/pipe should be named something like `Xtest-functional-server-server` to be consistent. Also if you are going to edit this file, see https://github.com/neovim/neovim/commit/64c7a36933baa21b3d635e9ee5d48c36541a91f9#commitcomment-15225023:

> I would write this as
> 
> ```Lua
>    it('sets v:servername _only_ on nvim startup unless all servers are stopped',
>    function()
> ```
> 
> . You really don’t want newline and indentation in a test name in output, it looks ugly.

**Comment by ZyX-I** [comment]
@sethjackson When I said “literally” I meant literally. `\` is a *valid* part of the name, *not* a path separator.

**Comment by sethjackson** [comment]
Oh right. Duh.

**Comment by ZyX-I** [comment]
@sethjackson On Unix (in POSIX filesystems, NTFS is the one BTW, though Windows API will not allow you to use NTFS like this) there are exactly *two* **bytes** that cannot be placed in a file name: forward slash (0x2F) and NUL (0x00): first is a path separator, second terminates string. Everything else can be in the file name.

**Comment by sethjackson** [comment]
@ZyX-I good to know. Thanks! :+1:

File: test/functional/server/server_spec.lua [hunk 1]
```diff
@@ -38,7 +38,7 @@ describe('serverstart(), serverstop()', function()
     eq('', eval('v:servername'))
 
     -- v:servername will take the next available server.
-    if eval('has("win32")') then
+    if eval('has("win32")') == 1 then
```
Comment (justinmk) [inline test/functional/server/server_spec.lua]: IMO this is a good reason to move this up into `helpers.os_is_win()`.

File: test/functional/server/server_spec.lua [hunk 2]
```diff
@@ -38,7 +38,7 @@ describe('serverstart(), serverstop()', function()
...
       nvim('command', "call serverstart('\\\\.\\pipe\\test_server_pipe')")
       eq('\\\\.\\pipe\\test_server_pipe', eval('v:servername'))
```
Comment (justinmk) [inline test/functional/server/server_spec.lua]: Using `[[ ]]` would avoid some escape madness here:

```lua
nvim('command', [[call serverstart('\\.\pipe\test_server_pipe')]])
eq('\\.\pipe\test_server_pipe', eval('v:servername'))
```

**Comment by justinmk** [comment]
Can you remove the newline from the `it()` description on line 20?

File: test/functional/server/server_spec.lua (hunk 2 above)
Comment (sethjackson) [inline test/functional/server/server_spec.lua]: I still think we need the extra escaping in the `eq` though if we use single quotes?
Although maybe we can use the brackets there too?

I'm not too familiar with Lua.

**Comment by sethjackson** [comment]
Newline removed.

File: test/functional/server/server_spec.lua [hunk 3]
```diff
@@ -17,8 +17,7 @@ describe('serverstart(), serverstop()', function()
     eq('', eval('$NVIM_LISTEN_ADDRESS'))
   end)
 
-  it([[sets v:servername _only_ on nvim startup
-      (unless all servers are stopped)]], function()
+  it([[sets v:servername _only_ on nvim startup (unless all servers are stopped)]], function()
```
Comment (ZyX-I) [inline test/functional/server/server_spec.lua]: My variant (additionally remove parenthesis, replace `[[]]` with `''` and move `function()` on the next line) keeps this line behind 80 characters cap.

File: test/functional/server/server_spec.lua [hunk 4]
```diff
@@ -38,8 +37,8 @@ describe('serverstart(), serverstop()', function()
     eq('', eval('v:servername'))
 
     -- v:servername will take the next available server.
-    if eval('has("win32")') then
-      nvim('command', "call serverstart('\\\\.\\pipe\\test_server_pipe')")
+    if eval('has("win32")') == 1 then
```
Comment (ZyX-I) [inline test/functional/server/server_spec.lua]: I would actually suggest writing this as

    local servername = (os_is_win() and [[\\.\pipe\Xtest-functional-server-server-pipe]]
                        or 'Xtest-functional-server-server-socket')
    funcs.serverstart(servername)
    eq(servername, meths.get_vvar('servername'))

. All unknown identifiers are from `helpers` package (one is missing, suggested by @justinmk).

File: test/functional/server/server_spec.lua (hunk 3 above)
Comment (sethjackson) [inline test/functional/server/server_spec.lua]: Ok updated.

File: test/functional/server/server_spec.lua [hunk 5]
```diff
@@ -17,8 +17,8 @@ describe('serverstart(), serverstop()', function()
     eq('', eval('$NVIM_LISTEN_ADDRESS'))
   end)
 
-  it([[sets v:servername _only_ on nvim startup
-      (unless all servers are stopped)]], function()
+  it('sets v:servername _only_ on nvim startup unless all servers are stopped)',
```
Comment (ZyX-I) [inline test/functional/server/server_spec.lua]: Changes to this file must not be part of the commit named “Windows: Fix api/vim_spec.lua functional test”.

File: test/functional/server/server_spec.lua [hunk 6]
```diff
@@ -38,12 +38,12 @@ describe('serverstart(), serverstop()', function()
     eq('', eval('v:servername'))
 
     -- v:servername will take the next available server.
-    if eval('has("win32")') then
-      nvim('command', "call serverstart('\\\\.\\pipe\\test_server_pipe')")
-      eq('\\\\.\\pipe\\test_server_pipe', eval('v:servername'))
+    if eval('has("win32")') == 1 then
+      nvim('command', [[call serverstart('\\.\pipe\Xtest-functional-server-pipe')]])
+      eq('\\\\.\\pipe\\Xtest-functional-server-pipe', eval('v:servername'))
```
Comment (ZyX-I) [inline test/functional/server/server_spec.lua]: server-server is not a typo. It is file server_spec.lua in server directory.

File: test/functional/server/server_spec.lua (hunk 2 above)
Comment (justinmk) [inline test/functional/server/server_spec.lua]: @sethjackson  Yeah sorry that was a typo, need `[[ ]]` in `eq()` as well.

Comment (sethjackson) [inline test/functional/server/server_spec.lua]: Ok. :)

**Comment by ZyX-I** [comment]
Ideally the file should look like

```Lua
local helpers = require('test.functional.helpers')
local eq, neq, eval = helpers.eq, helpers.neq, helpers.eval
local clear, meths, funcs = helpers.clear, helpers.meths, helpers.funcs
local os_is_win, exc_exec = helpers.os_is_win, helpers.exc_exec

describe('serverstart(), serverstop()', function()
  before_each(clear)

  it('sets $NVIM_LISTEN_ADDRESS on first invocation', function()
    -- Unset $NVIM_LISTEN_ADDRESS
    meths.command('let $NVIM_LISTEN_ADDRESS = ""')

    local s = funcs.serverstart()
    assert(s ~= nil and s:len() > 0, "serverstart() returned empty")
    eq(s, eval('$NVIM_LISTEN_ADDRESS'))
    funcs.serverstop(s)
    eq('', eval('$NVIM_LISTEN_ADDRESS'))
  end)

  it('sets v:servername _only_ on nvim startup unless all servers are stopped)',
  function()
    local initial_server = meths.get_vvar('servername')
    assert(initial_server ~= nil and initial_server:len() > 0,
           'v:servername was not initialized')

    -- v:servername is readonly so we cannot unset it--but we can test that it
    -- does not get set again thereafter.
    local s = funcs.serverstart()
    assert(s ~= nil and s:len() > 0, "serverstart() returned empty")
    neq(initial_server, s)

    -- serverstop() does _not_ modify v:servername...
    funcs.serverstop(s)
    eq(initial_server, meths.get_vvar('servername'))

    -- ...unless we stop _all_ servers.
    funcs.serverstop(funcs.serverlist()[1])
    eq('', meths.get_vvar('servername'))

    -- v:servername will take the next available server.
    local servername = (os_is_win()
                        and [[\\.\pipe\Xtest-functional-server-server-pipe]]
                        or 'Xtest-functional-server-server-socket')
    funcs.serverstart(servername)
    eq(servername, meths.get_vvar('servername'))
  end)

  it('serverstop() ignores invalid input', function()
    eq(0, exc_exec('call serverstop("")'))
    eq(0, exc_exec('call serverstop("bogus-socket-name")'))
  end)
end)

describe('serverlist()', function()
  before_each(clear)

  it('returns the list of servers', function()
    -- There should already be at least one server.
    local n = #funcs.serverlist()

    -- Add a few
    local servs = {'should-not-exist', 'another-one-that-shouldnt'}
    for _, s in ipairs(servs) do
      eq(s, funcs.serverstart(s))
    end

    local new_servs = funcs.serverlist()

    -- Exactly #servs servers should be added.
    eq(n + #servs, #new_servs)
    -- The new servers should be at the end of the list.
    for i = 1, #servs do
      eq(servs[i], new_servs[i + n])
      funcs.serverstop(servs[i])
    end
    -- After calling serverstop() on the new servers, they should no longer be
    -- in the list.
    eq(n, #funcs.serverlist())
  end)
end)
```

. Though this is rather out of the scope.

**Comment by sethjackson** [comment]
Yeah. I think we have other tests that aren't using the helpers and using things like `call` directly IIRC.

Ok to not cleanup this entire file in this PR?

**Comment by ZyX-I** [comment]
@sethjackson It would be not OK to clean up the file. And you need either two commits or different commit comment.

**Comment by sethjackson** [comment]
Ok. What would a good commit message be?

File: test/functional/server/server_spec.lua [hunk 7]
```diff
@@ -17,34 +18,32 @@ describe('serverstart(), serverstop()', function()
     eq('', eval('$NVIM_LISTEN_ADDRESS'))
   end)
 
-  it([[sets v:servername _only_ on nvim startup
-      (unless all servers are stopped)]], function()
-    local initial_server = eval('v:servername')
+  it('sets v:servername _only_ on nvim startup unless all servers are stopped',
+  function()
+    local initial_server = meths.get_vvar('servername')
```
Comment (justinmk) [inline test/functional/server/server_spec.lua]: This is not preferable IMO. Vimscript is more recognizable (for most Vim devs hopefully).

Comment (ZyX-I) [inline test/functional/server/server_spec.lua]: This variant does not invoke heavier parsing. Also vim_get_vvar API call is less used and it is better to indirectly test it.

Comment (sethjackson) [inline test/functional/server/server_spec.lua]: Whatever we do I would prefer to be consistent in a test file.

If `vim_get_vvar` is not tested well enough we should just add more tests for it IMO.

File: test/functional/server/server_spec.lua [hunk 8]
```diff
@@ -17,34 +18,32 @@ describe('serverstart(), serverstop()', function()
...
     -- does not get set again thereafter.
-    local s = eval('serverstart()')
+    local s = funcs.serverstart()
     assert(s ~= nil and s:len() > 0, "serverstart() returned empty")
     neq(initial_server, s)
 
     -- serverstop() does _not_ modify v:servername...
-    nvim('command', "call serverstop('"..s.."')")
-    eq(initial_server, eval('v:servername'))
+    funcs.serverstop(s)
```
Comment (justinmk) [inline test/functional/server/server_spec.lua]: OTOH this is definitely an improvement.

File: test/functional/server/server_spec.lua (hunk 7 above)
Comment (ZyX-I) [inline test/functional/server/server_spec.lua]: I personally tend to prefer the API functions (when I know about them) because API is *specific*. For test purposes this is not an advantage though.
